satellite/satellitedb/overlaycache.go

```diff
@@ -60,7 +62,23 @@ func (cache *overlaycache) SelectStorageNodes(ctx context.Context, count int, cr
 		args = append(args, v.Major, v.Major, v.Minor, v.Minor, v.Patch)
 	}
 
-	return cache.queryFilteredNodes(ctx, criteria.Excluded, count, safeQuery, args...)
+	var totalNodes []*pb.Node
+	for i := 0; i < 3; i++ {
```

Comment:
why do we need to iterate 3 times? If we pass `count` into `cache.queryFilteredNodes`, it should return that many nodes, and if it doesn't, calling it again won't return more nodes, right?

Reply:
So the subquery gets us a number of reputable nodes equal to count. We then SELECT DISTINCT ON last_ip from that set of reputable nodes to filter out duplicate IPs. If we filtered any entries out, the number of nodes < count

Reply:
I agree with @mobyvb here, we should try to construct the query differently then.

Reply:
This query is about twice as fast as the old one, `Row_number() OVER(PARTITION BY last_ip ORDER BY RANDOM()`. @jtolds preferred speed with potential need for retrying

Reply:
We could also increase `count` by some percentage to offset the possibility of duplicates requiring a subsequent query?

Reply:
@egonelbre Objections?

Reply:
heh, cam has evaluated every possible option for this query i can think of, learning deep, dark postgres-specific secrets in the meantime. this query in the common case is much faster, and the potential for needing retries is rare. we have not been able to find another solution that provides this level of overall speed. my vote is we continue forward, unless someone can suggest a faster query that doesn't have this problem. there hasn't been a completely perfect solution found so far but this one seems to have the least downsides believe it or not

Reply:
Egon suggested to move this out of the DB and do it with in RAM maps. I agree with him on that.
This would remove the need for retries to optimize performance, right?

Reply:
Would that eliminate the need for retries? If last_ip isn't in the DB, how can we know that there are no duplicates in the set we retrieve? Or on the other hand, we have a map of nodeIDs and their last IPs; we choose x nodeIDs with unique IPs from that list and query for the reputable ones from that list. What if not enough of them are reputable?

Reply:
The more nodes we have the less likely we will have the need for the extra nodes... however it might be still useful to always query some amount of more.

I think the more performant solution would be to keep everything in memory and using combiner queue flush things to database... depending on the query make it wait for sync to database. In essence, yes, it would remove the need for retries because we can write whatever code we like.

Reply:
Is that something you would like done in this PR, or should we make another ticket?

Reply:
Let's make another ticket for that suggestion

Reply:
@cam-a Lets get your PR merged with this approach and we monitor the overhead and performance with it.
As Jen already stated, lets make a new ticket/PR for the additional node request.